src/backend/distributed/connection/remote_commands.c

```diff
@@ -75,6 +75,90 @@ ForgetResults(MultiConnection *connection)
 
 
 /*
+ * NonblockingForgetResults clears a connection from pending activity if doing
+ * so does not require network IO. Returns true if successful, false
+ * otherwise.
+ */
+bool
+NonblockingForgetResults(MultiConnection *connection)
+{
+	PGconn *pgConn = connection->pgConn;
+	bool wasNonblocking = PQisnonblocking(pgConn);
+
+	if (PQstatus(pgConn) != CONNECTION_OK)
+	{
+		return false;
+	}
+
+	while (true)
+	{
+		PGresult *result = NULL;
+
+		CHECK_FOR_INTERRUPTS();
+
+		if (!wasNonblocking)
```

Comment:
These annoy me a bit, since we now have this set / unset in 7 different places. Ideally, we would just set our sockets to non-blocking mode, though I'm guessing that will mess up copy. To simplify the code, could we maybe add a flag in `MultiConnection` that says whether the connection is blocking so we can get rid of the `wasNonBlocking` checks? That would also simplify asynchronous commit.